packages/apps/patterns/react-appkit/src/components/CrudList/CrudList.tsx

```diff
@@ -0,0 +1,226 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import { useControllableState } from '@radix-ui/react-use-controllable-state';
+import React, { ChangeEvent, ComponentPropsWithoutRef, forwardRef, KeyboardEvent, ReactNode, useCallback } from 'react';
+
+import {
+  List,
+  ListItem,
+  DragEndEvent,
+  arrayMove,
+  ListItemHeading,
+  Input,
+  ListItemEndcap,
+  ListItemDragHandle,
+  ListProps,
+  InputProps,
+  ListItemProps,
+  mx
+} from '@dxos/react-components';
+
+export interface CrudListItemSlots {
+  listItem?: ListItemProps['slots'];
+  input?: InputProps['slots'];
+}
+
+export interface CrudListItemProps {
+  id: string;
+  defaultCompleted?: boolean;
+  completed?: boolean;
+  onCompletedChange?: (completed: boolean) => void;
+  defaultTitle?: string;
+  title?: string;
+  onTitleChange?: (event: ChangeEvent<HTMLInputElement>) => void;
+  slots?: CrudListItemSlots;
+}
+
+export interface CrudListSlots {
+  root?: Omit<ComponentPropsWithoutRef<'div'>, 'children'>;
+  listHeading?: InputProps['slots'];
+  list?: ListProps['slots'];
+  addItem?: InputProps['slots'];
+}
+
+export interface CrudListProps {
+  id: string;
+  labelId: string;
+  defaultNextItemTitle?: string;
+  nextItemTitle?: string;
+  onNextItemTitleChange?: (event: ChangeEvent<HTMLInputElement>) => void;
+  defaultItemIdOrder?: string[];
+  itemIdOrder?: string[];
+  onItemIdOrderChange?: (itemIdOrder: string[]) => void;
+  children?: ReactNode;
+  slots?: CrudListSlots;
+}
+
+export const useCrudListKeyboardInteractions = (hostId: string) => {
+  const hostAttrs = { 'data-focus-series-host': hostId };
+  const itemAttrs = { 'data-focus-series': 'crudList' };
+  const onListItemInputKeyDown = useCallback(
+    (event: KeyboardEvent<Element>) => {
+      const inputsInScope = Array.from(
+        document.querySelectorAll<HTMLInputElement>(
+          `[data-focus-series-host="${hostId}"] [data-focus-series="crudList"]`
+        )
+      );
+      const targetIndex = inputsInScope.findIndex((sibling) => sibling === event.target);
+      switch (event.key) {
+        case 'Enter':
+        case 'PageDown': {
+          if (targetIndex < inputsInScope.length - 1) {
+            event.preventDefault();
+            inputsInScope[targetIndex + 1].focus();
+          }
+          break;
+        }
+        case 'PageUp': {
+          if (targetIndex > 0) {
+            event.preventDefault();
+            inputsInScope[targetIndex - 1].focus();
+          }
+          break;
+        }
+      }
+    },
+    [hostId]
+  );
+  return { onListItemInputKeyDown, hostAttrs, itemAttrs };
+};
+
+export const CrudList = ({
+  children,
+  labelId,
+  nextItemTitle,
+  defaultNextItemTitle,
+  onNextItemTitleChange,
+  defaultItemIdOrder,
+  itemIdOrder,
+  onItemIdOrderChange,
```

Comment:
nice to group params in order: required, optional, callbacks (at least group callbacks)

Reply:
I’m not grouping callbacks here because e.g. `itemIdOrder` comes in three since it is a controllable state:

- `defaultItemIdOrder` (uncontrolled value)
- `itemIdOrder` (controlled value)
- `onChangeItemIdOrder` (callback)

I feel these props should come together so it's easier to be sure I got all three for each prop that is part of a controllable state.